File: golang/speedtest/fronting.go
package speedtest

import (
	utils "CFScanner/utils"
	"crypto/tls"
	"fmt"
	"io"
	"log"
	"net"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// FrontingTest conducts a fronting test on an ip and return true if status 200 is received
func FrontingTest(ip string, proxies map[string]string, timeout time.Duration) bool {

	var success = false

	// set proxy to request
	var proxy *url.URL
	// set proxies to req header
	for _, v := range proxies {
		proxy, _ = url.Parse(v)
	}

	compatibleIP := ip
	if strings.Contains(ip, ":") {
		compatibleIP = fmt.Sprintf("[%s]", ip)
	} else {
		compatibleIP = ip
	}

	req, err := http.NewRequest("GET", fmt.Sprintf("https://%s", compatibleIP), nil)
	if err != nil {
		fmt.Printf("Error creating request for IP %s: %v\n", ip, err)
		return false
	}
	req.Host = "speed.cloudflare.com"

	client := &http.Client{
		Timeout: timeout * time.Second,
		Transport: &http.Transport{Proxy: http.ProxyURL(proxy),
			TLSClientConfig: &tls.Config{
				ServerName:         "speed.cloudflare.com",
				InsecureSkipVerify: true,
			},
		},
	}

	resp, err := client.Do(req)

	if err != nil {
		switch err := err.(type) {
		case net.Error:
			if err.Timeout() {
				log.Printf("%vFAIL%v %v%15s Fronting test connect timeout%v\n",
					utils.Colors.FAIL, utils.Colors.ENDC, utils.Colors.WARNING, ip, utils.Colors.ENDC)
			} else {
				log.Printf("%vFAIL%v %v%15s Fronting test connection error: %v%v\n",
					utils.Colors.FAIL, utils.Colors.ENDC, utils.Colors.WARNING, ip, err, utils.Colors.ENDC)
			}
		default:
			fmt.Printf("%vFAIL%v %v%15s Fronting test unknown error: %v%v\n",
				utils.Colors.FAIL, utils.Colors.ENDC, utils.Colors.WARNING, ip, err, utils.Colors.ENDC)
		}
		return false
	}

	defer func(Body io.ReadCloser) {
		err := Body.Close()
		if err != nil {
			fmt.Errorf("error occured when closing fronting body %v", err)
		}
	}(resp.Body)

	if resp.StatusCode != http.StatusOK {
		log.Printf("%vFAIL%v %v%s Fronting test error : %d%v\n",
			utils.Colors.FAIL, utils.Colors.ENDC, utils.Colors.WARNING, ip, resp.StatusCode, utils.Colors.ENDC)
	} else {
		success = true
	}

	return success
}

File: python/speedtest/fronting.py
import requests

from report.clog import CLogger

logger = CLogger("fronting")


def fronting_test(
    ip: str,
    timeout: float
) -> bool:
    """conducts a fronting test on an ip and return true if status 200 is received

    Args:
        ip (str): ip for testing
        timeout (float): the timeout to wait for ``requests.get`` result

    Returns:
        bool: True if ``status_code`` is 200, False otherwise
    """
    s = requests.Session()
    s.get_adapter(
        'https://').poolmanager.connection_pool_kw['server_hostname'] = "speed.cloudflare.com"
    s.get_adapter(
        'https://').poolmanager.connection_pool_kw['assert_hostname'] = "speed.cloudflare.com"

    try:
        compatible_ip = f"[{ip}]" if ":" in ip else ip
        r = s.get(
            f"https://{compatible_ip}",
            timeout=timeout,
            headers={"Host": "speed.cloudflare.com"}
        )
        if r.status_code != 200:
            return f"[red]NO[/red] [dark_orange3]{ip:15s}[/dark_orange3][yellow] fronting error {r.status_code} [/yellow]"
        else:
            success = True
    except requests.exceptions.ConnectTimeout as e:
        return f"[red]NO[/red] [dark_orange3]{ip:15s}[/dark_orange3][yellow] fronting connect timeout[/yellow]"
    except requests.exceptions.ReadTimeout as e:
        return f"[red]NO[/red] [dark_orange3]{ip:15s}[/dark_orange3][yellow] fronting read timeout[/yellow]"
    except requests.exceptions.ConnectionError as e:
        return f"[red]NO[/red] [dark_orange3]{ip:15s}[/dark_orange3][yellow] fronting connection error[/yellow]"
    except Exception as e:
        logger.error(f"Fronting test Unknown error {ip:15}")
        logger.exception(e)
        return f"[red]NO[/red] [dark_orange3]{ip:15s}[/dark_orange3][yellow] fronting Unknown error[/yellow]"
    

    return "OK"
